From: Gerald Schaefer <gerald.schaefer@de.ibm.com>
Subject: s390/vmlogrdr: fix IUCV buffer allocation
Patch-mainline: v4.10-rc1
Git-commit: 5457e03de918f7a3e294eb9d26a608ab8a579976
References: bnc#1025702, LTC#152144

Description:  s390/vmlogrdr: fix IUCV buffer allocation
Symptom:      Addressing exception or memory corruption.
Problem:      The buffer for iucv_message_receive() needs to be below 2 GB. In
              __iucv_message_receive(), the buffer address is casted to an u32.
              Depending on the address returned from get_zeroed_page(), this
              would result in either memory corruption or an addressing
              exception when using addresses >= 2 GB.
Solution:     Use GFP_DMA for the buffer allocation.
Reproduction: Load vmlogrdr module (and hit the case where get_zeroed_page()
              returns an address >= 2 GB)

Upstream-Description:

              s390/vmlogrdr: fix IUCV buffer allocation

              The buffer for iucv_message_receive() needs to be below 2 GB. In
              __iucv_message_receive(), the buffer address is casted to an u32, which
              would result in either memory corruption or an addressing exception when
              using addresses >= 2 GB.

              Fix this by using GFP_DMA for the buffer allocation.

              Cc: stable@vger.kernel.org
              Signed-off-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.com>
---
 drivers/s390/char/vmlogrdr.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/s390/char/vmlogrdr.c
+++ b/drivers/s390/char/vmlogrdr.c
@@ -876,7 +876,7 @@ static int __init vmlogrdr_init(void)
 		goto cleanup;
 
 	for (i=0; i < MAXMINOR; ++i ) {
-		sys_ser[i].buffer = (char *) get_zeroed_page(GFP_KERNEL);
+		sys_ser[i].buffer = (char *) get_zeroed_page(GFP_KERNEL | GFP_DMA);
 		if (!sys_ser[i].buffer) {
 			rc = -ENOMEM;
 			break;
